python_modules/dagster-externals/dagster_externals_tests/test_external_execution.py

```diff
@@ -35,11 +41,50 @@ def temp_script(script_fn: Callable[[], Any]) -> Iterator[str]:
         yield file.name
 
 
-def test_external_subprocess_asset(capsys):
+@pytest.mark.parametrize(
+    ("context_source_spec", "message_sink_spec"),
+    [
+        ("default", "default"),
+        ("default", "user/file"),
+        ("user/file", "default"),
+        ("user/file", "user/file"),
+        ("user/env", "default"),
+        ("user/env", "user/file"),
+    ],
+)
+def test_external_subprocess_asset(capsys, tmpdir, context_source_spec, message_sink_spec):
+    if context_source_spec == "default":
+        context_source = None
+    elif context_source_spec == "user/file":
+        context_source = get_file_context_source(os.path.join(tmpdir, "input"))
+    elif context_source_spec == "user/env":
+        context_source = get_env_context_source()
+    else:
+        assert False, "Unreachable"
+
+    if message_sink_spec == "default":
+        message_sink = None
+    elif message_sink_spec == "user/file":
+        message_sink = get_file_message_sink(os.path.join(tmpdir, "output"))
+    else:
+        assert False, "Unreachable"
+
     def script_fn():
-        from dagster_externals import ExternalExecutionContext, init_dagster_externals
+        import os
 
-        init_dagster_externals()
+        from dagster_externals import (
+            ExternalExecutionContext,
+            ExternalExecutionEnvContextSource,
+            init_dagster_externals,
+        )
+
+        context_source = (
+            ExternalExecutionEnvContextSource()
+            if os.getenv("CONTEXT_SOURCE_SPEC") == "user/env"
+            else None
+        )
```

Comment:
I ended up doing similar in https://github.com/dagster-io/dagster/pull/16122/files for stdtout/err message sink, but it also seems like it makes sense to support these in the default source/sink. With the existing env var already a dict, i half way implemented a pattern like dagster config Selector where you expect a single top level key but the key changes the behavior :
* context source
  * `path` : file at this path
  * `context_data`: its right here in the json
* message sink
  * `path`: file at this path
  * `sys`: value is either `stderr` /  `stdout` , write to there 
  
Could imagine extrapolating to a set-up where users provide a dictionary of there special key to handler for that key on the external side. This gets a set-up where users add support past the defaults instead of replacing the defaults. 

Not strong set on this, just floating the idea

Reply:
https://gist.github.com/alangenfeld/60db1fb7d5df288c4b221fb723b2955d